### ISSUE TITLE
Create OUI React components

### ISSUE THREAD

**Issue by danoc** [optimizely/oui#378]
Optimizely will soon be introducing React to the front end. It would be great to have a collection of OUI React components ready to go for engineers to use on day one. This will greatly increase UI consistency and improve developer productivity.

Here's a brain-dump of the work required to create these components. The steps are ordered but they can mostly be done in parallel. Please leave feedback—especially if any of this sounds wonky!

### Step 1: Identify components and properties

We should identify the components that we'd like to create. Although we have a "Components" section in our documentation, some of the items would be split into smaller components. `_forms.scss`, for example, contains code for text inputs, textareas, radio buttons, and a number of other potential components. (The resources at the bottom of this post have great example components.)

I've started identifying our components in a [JSON file within the `daniel/patterns.json`](https://github.com/optimizely/oui/blob/daniel/patterns.json/patterns.json) branch. This `patterns.json` file  is will serve source of truth (at least for now) for the list of components and their properties.

![screenshot 2016-04-18 10 55 05](https://cloud.githubusercontent.com/assets/916038/14614016/093fa792-0554-11e6-8e6c-f7aea1051f82.png)

I'd like for the React components to eventually pull component data and "rules" from `patterns.json`. The rules would map to props and propTypes. This is a step towards making our components language and framework agnostic and has two main benefits:
 
1. The `patterns.json` file can be used to generate an OUI React documentation site, keep design.optimizely.com up to date, and power React component propTypes. It can also be used on non-React websites (even static websites) to enforce that patterns are being used appropriately and throw client-side console errors as React does with props.
2. It becomes easier for non-technical designers to view and potentially edit the patterns.

**Update (June 27, 2016):** I've tabled the idea of a `patterns.json` file but would like to revisit it in the future.

### Step 2: Set up React development environment within OUI

I've added React, webpack, babel, and some other tools in the [`daniel/react`](https://github.com/optimizely/oui/tree/daniel/react) branch. You can currently checkout the branch and run `npm run react:serve`, but it doesn't yet watch files for changes.

**Update (March 18, 2016):** I've updated the build tools to fix the development environment issues. `npm run react:serve` will now both watch the files, re-run webpack, and reload browser-sync.

### Step 3: Move towards a componentized directory structure

This is described in issue #368 and depends on identifying the components. We can merge this before completing the other steps.

### Step 4: Build the stateless React components

I've created a few [sample React components](https://github.com/optimizely/oui/tree/daniel/react/src/js/react), but we should get them code reviewed and look at other React UI libraries to ensure that we're on the right path.

We can prioritize these components before building and create pull requests and merge into a feature branch.

Let's reference the `patterns.json` file when building the components and decide later if it is worth connecting the JSX to the JSON.

### Future Ideas
- We should definitely create great, automatically generated, documentation. [Rebass](http://jxnblk.com/rebass/) is an example of a library that does this (they mentioned the tools at the bottom).
- Standardize our prototyping environment and process. A tool like [React Storybook](https://github.com/kadirahq/react-storybook) could be helpful. @daverau-optimizely also has an idea for a sweet component playground/explorer view.

***

### Resources:
* Rebass: http://jxnblk.com/rebass/
* React-Bootstrap: https://react-bootstrap.github.io/
* Reactstrap: https://reactstrap.github.io/
* Mesosphere UI Components: https://mesosphere.github.io/reactjs-components/
* MUI (Material UI) React: https://www.muicss.com/docs/v1/react/introduction
* Pivotal: https://styleguide.cfapps.io/

***
cc: @kwalker3690, @daverau-optimizely, @tscanlin

**Comment by danoc** [comment]
I've updated the branch `daniel/react` to properly watch JSX files, rebuild, and refresh browser-sync with the `npm run webpack-watch` command.

**Comment by tscanlin** [comment]
@danoc great work laying out a solid plan to get us going on this!

I'm curious if you had considered keeping the react components in a different repo than oui? It seems that keeping components in another repo would keep the jsx decoupled from our css which is probably a good thing.

Also, I like the patterns.json idea but I definitely think we should invest the time now to have the jsx files reference it so we don't duplicate logic / create complexity.

Finally, thanks for including the links to the other react component libraries, I hadn't heard of some of those! What do you guys think about spending some time to evaluate these different component libraries and see what we like / dislike about each one so we can setup some goals for this project. I also think it could be worth trying to borrow some of the components from one of these to get us a head start on this. Thoughts?

**Comment by danoc** [comment]
Hey @tscanlin! I thought about the separate repository approach for a while and created a dependency graph to represent it:

![oui-oui react-optimizely dependency graph 1](https://cloud.githubusercontent.com/assets/916038/14684467/ec226e3c-06e4-11e6-858b-d0679f0f932c.png)

This would cause headaches when OUI releases breaking SCSS changes. If OUI releases `v12.0.0` then the OUI-React repo must be updated to support `v12.0.0` _before_ the Optimizely app. If we don't follow this rule and the Optimizely app is updated to support `v12.0.0` without modifying OUI-React, then we can end up in a state where the JSX relies on a class that was removed in the version of OUI.

I'd also argue that the industry is moving towards a world where HTML/CSS/JS are tightly coupled (CSS modules, Web Components, etc...).

I met with Jordan yesterday to do an in-person code review of the existing components, but I totally agree that we should research the other libraries a bit more. I'll look through them a bit more today and see what [we're currently doing different](https://github.com/optimizely/oui/tree/daniel/react/src/js/react). I'd recommend doing the same. Feel free to put some time on my calendar or post here with findings!

**Comment by danoc** [comment]
@tscanlin – Here are some notes from looking at the ones I linked to above.

**Rebass:**
* Inlines component CSS and uses [React Context](https://facebook.github.io/react/docs/context.html) to pass styles from the parent to the child. This works similarly to global variables.
* They discourage styling with CSS and have no CSS dependencies.
* They use `react-addons-test-utils` and `expect` for testing.
* Documentation is automatically generated using third party tools and very developer friendly

**React-Bootstrap**
* They don’t include Bootstrap CSS but require loading it for the components to work. Their “Getting Started” page just has people point to the latest version of Bootstrap.
* They use classes instead of pure functions

**reactstrap**
* Use a neat tool called [Tether](http://tether.io/) for positioning dropdowns, tooltips, and popovers.
* Seems less mature than the others

**mesosphere**
* Includes their SCSS as a dependency and lists it in the peerDependencies to ensure that the correct version is being used

**Pivotal UI**
* [Have tabs in their navigation](https://styleguide.cfapps.io/) for React or CSS
* Package each of their components as separate NPM modules but all of the modules live within the same repo and [have their own package.json files](https://github.com/pivotal-cf/pivotal-ui/blob/5e822132192a1448a6bdcae0a625388c7efe73b0/library/src/pivotal-ui-react/labels/package.json)
* Interesting documentation about their development and release process: https://github.com/pivotal-cf/pivotal-ui/blob/development/CORE_TEAM_DOCS.md
* The React components [require the component CSS at the top of each file](https://github.com/pivotal-cf/pivotal-ui/blob/bfad6beb1b923b8e3b75e0f12368881d7461b317/library/src/pivotal-ui-react/labels/labels.js#L3)

**Comment by tscanlin** [comment]
@danoc I agree that the dependency tree could get complicated and oui-react would need to be updated before Optimizely in the case of a breaking change, that is totally true.

I guess I was imagining that Optimizely wouldn't actually necessarily need to have the oui dependency if oui-react has it. But there probably is some extra complexity in making them separate and I'm cool with keeping them together as well. This decision might result in more major version bumps for oui, but that's probably ok.

I really like your summaries of the other react component libraries!

Rebass and Pivotal UI seem to be the most extensive / mature at this point, what do you think?

Here's a list of some objectives we should consider for our library. Feel free to comment or add more!
- self-contained component directory structure. (css, html, jsx, tests, docs)
- auto generated documentation from component docs (or maybe code comments)
- interactive documentation would be awesome (playground?)
- right now I favor keeping css as it is rather than putting into js / inlining (anyone wanna convince me otherwise? :) )
- uses existing oui classes + any new classes that are necessary
- jest would be awesome for testing, its super fast and uses jsdom for rendering components.

Would scribe-sass work for generating docs?

**Comment by danoc** [comment]
@tscanlin – Yeah, those two are definitely my favorites. I think Rebass is most similar to what we're trying to do. Pivotal UI is incredibly advanced and I think we can learn some things from it.

+1 to all the other comments. ScribeSass would not work in its current state. I have to think about that a bit more.

**Comment by tscanlin** [comment]
Wow, Origami is really advanced! Both in its breadth, but also the level of detailed info for each component. Very cool :)

As for ScribeSass, we may want to consider just using markdown for docs / examples since they are probably easier to work with and if we have the componentized folder structure keeping it all in the sass file is probably less necessary. But I'm open to discussing this.

**Comment by danoc** [comment]
@daverau-optimizely has some good ideas for neat ways to display examples!

**Comment by tscanlin** [comment]
Feel free to edit and comment on this doc from our meeting:
https://docs.google.com/document/d/1SitU0fJnDIWmyFe_pr0Jc6b94A7E3aZgSFk4aDEQBu0/edit#
